pal/GCC/LINUX/ulib_port.h

```diff
@@ -25,41 +25,60 @@
 #define COUNT_TYPE uint32_t
 #elif defined(AZURE_ULIB_C_USE_STD_ATOMIC)
 #define COUNT_TYPE _Atomic uint32_t
-#else  // AZURE_ULIB_C_USE_GNU_C_ATOMIC
+#else // AZURE_ULIB_C_USE_GNU_C_ATOMIC
 #define COUNT_TYPE uint32_t
 #endif // defined(AZURE_ULIB_C_ATOMIC_DONTCARE)
 
-
 /*the following macros increment/decrement a ref count in an atomic way, depending on the platform*/
 /*The following mechanisms are considered in this order
 AZURE_ULIB_C_ATOMIC_DONTCARE does not use atomic operations
 - will result in ++/-- used for increment/decrement.
 C11
 - will result in #include <stdatomic.h>
 - will use atomic_fetch_add/sub;
-- about the return value: "Atomically, the value pointed to by object immediately before the effects"
-gcc
+- about the return value: "Atomically, the value pointed to by object immediately before the
+  effects" gcc
 - will result in no include (for gcc these are intrinsics build in)
 - will use __sync_fetch_and_add/sub
-- about the return value: "... returns the value that had previously been in memory." (https://gcc.gnu.org/onlinedocs/gcc-4.4.3/gcc/Atomic-Builtins.html#Atomic-Builtins)
+- about the return value: "... returns the value that had previously been in memory."
+  (https://gcc.gnu.org/onlinedocs/gcc-4.4.3/gcc/Atomic-Builtins.html#Atomic-Builtins)
 */
 
-
 #if defined(AZURE_ULIB_C_ATOMIC_DONTCARE)
-#define AZ_ULIB_PORT_ATOMIC_INC_W(count) ++(*count)
-#define AZ_ULIB_PORT_ATOMIC_DEC_W(count) --(*count)
+#define AZ_ULIB_PORT_ATOMIC_INC_W(count) ++(*(count))
+#define AZ_ULIB_PORT_ATOMIC_DEC_W(count) --(*(count))
+static inline uint32_t AZ_ULIB_PORT_ATOMIC_EXCHANGE_W(volatile uint32_t* addr, uint32_t val) {
+  uint32_t prev = *addr;
+  *addr = val;
+  return prev;
+}
+static inline uint32_t AZ_ULIB_PORT_ATOMIC_EXCHANGE_PTR(volatile void** addr, void* val) {
+  void* prev = *addr;
+  *addr = val;
+  return prev;
+}
 
 #elif defined(AZURE_ULIB_C_USE_STD_ATOMIC)
 #include <stdatomic.h>
```

Comment:
do we need to worry about C++ here and have `<atomic>` in that case

Reply:
Nobody complains about it, but sounds a good idea. 

---
In reply to: [343374309](https://github.com/Azure/azure-ulib-c/pull/47#discussion_r343374309) [](ancestors = 343374309)